### PR TITLE
Trim analyzer: Implement intrinsics handling of object.GetType

### LINKED ISSUE #86921: Trim analyzer doesn't warn on `GetType().GetMethods()`
The following doesn't seem to produce an analyzer warning:

```C#
class C
{
    public static void Main()
    {
        M("abc");
    }
    public static void M(object o)
    {
        var methods = o.GetType().GetMethods();
    }
}
```

### PR DESCRIPTION
**Pull request by vitek-karas**
This fixes https://github.com/dotnet/runtime/issues/86921.

Analyzer so far didn't handle correct data flow around `object.GetType` and `DynamicallyAccessedMembersAttribute` on types. This change implements that behavior.

Main changes:
* Move `IValueWithStaticType` to the shared code and refactor its existing usage in trimmer/AOT to use the shared code instead. Also implement it for the analyzer.
* Refactor method call handling in the analyzer to a single static method which is called both from the visitor and from the patterns.
* In order to get same behavior, start tracking values for all fields and method parameters.

Outside of the actual fix, the other main change is that analyzer now tracks values for all fields and method parameters, regardless if their type is interesting to analysis. This is necessary because the static type now matters, even if it's something else than `System.Type`.

The effect of that is that the analyzer now recognizes lot more invalid cases because it can determine if the value is something unrecognizable. Before the change such values where tracked as "empty", and thus analysis ignored them. Now they're track as "value of a field, without annotations" which can lead to producing more warnings.

That means this resolves lot of cases in https://github.com/dotnet/linker/issues/2755. At least all the test cases which were added because of that bug, or which expected different behavior because of it now produce consistent behavior with trimmer/NativeAOT.

### CODE REVIEW

**Comment by vitek-karas** [comment]
Tracking more values means that the analyzer is potentially slower. I'll try to measure this somehow.

**Comment by agocke** [comment]
This looks good -- but I'm not sure why some things weren't being tracked before.

That is, my understanding of why `typeof(Blah).MakeGenericType` doesn't warn is that the receiver wasn't tracked for some reason. But `typeof(Blah)` is definitely an instance of `System.Type`. So why wasn't it tracked?

**Review by sbomer — APPROVED** [review]
One question about the treatment of "empty", otherwise LGTM, thanks!

File: src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisVisitor.cs [hunk 1]
```diff
@@ -296,6 +270,108 @@ public override MultiValue HandleMethodCall (IMethodSymbol calledMethod, MultiVa
...
+			case IntrinsicId.TypeDelegator_Ctor:
+				if (operation is IObjectCreationOperation)
+					AddReturnValue (arguments[0]);
+
+				break;
+
+			case IntrinsicId.Object_GetType: {
+					// In the analyzer if the instance is an empty set of values, it actually means "unknown" value
+					// which in this case we want to warn about.
+					MultiValue instanceValue = instance.IsEmpty () ? UnknownValue.Instance : instance;
```
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisVisitor.cs]: Good point - this is basically a leftover from before I changed the source of the values. Removed this and it only "broke" two rather edge case tests.

**Pull request merged**